### PR TITLE
Hclang 140

### PR DESCRIPTION
**Pull request by technight**
### Description
Added a section about the generated c++ code

### Issues solved 
add these with the respective key as HCLANG-1 (as an example)
 - HCLANG-140

### Issues opened as a result of this 
Add something here if new issues arise because of this. For instance if new features-requests or bugs arise from this PR. If this is not the case, then remove this section.

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/4.Solution/GeneratedCppCode.tex [hunk 1]
```diff
@@ -0,0 +1,87 @@
+% !TeX root = ../main.tex
+
+\subsection{Generated C++ code}
+
+While the previous section primarily focused on the code generation from an abstract point of view, this section will go into depth about the generated C++ code.
+Since C++ is the output language of the HCL compiler, all the language concepts of HCL has to somehow be represented in valid C++ code.
+
+\textbf{Primitives}
+
+The primitives of HCL includes numbers, text and booleans.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: (Maybe) Change "The primitives" -> "The primitive types"

File: report/4.Solution/GeneratedCppCode.tex [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+\subsection{Generated C++ code}
+
+While the previous section primarily focused on the code generation from an abstract point of view, this section will go into depth about the generated C++ code.
+Since C++ is the output language of the HCL compiler, all the language concepts of HCL has to somehow be represented in valid C++ code.
+
+\textbf{Primitives}
+
+The primitives of HCL includes numbers, text and booleans. 
+Booleans are straightforward, as HCL's boolean behaves exactly like the C++ boolean type.
+This means that the regular boolean type can be used for output type of booleans.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "used for output" -> "used as the output"

File: report/4.Solution/GeneratedCppCode.tex [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+As HCL aims for simplicity, only one number type was introduced. 
+However this type should also be able to store non-integers. 
+Therefore the simple solution was to turn all HCL numbers into C++'s double type.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "to turn all" -> "to format all"

File: report/4.Solution/GeneratedCppCode.tex [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+The solution with doubles is not ideal, but it will fulfill the requirements of the number type in HCL.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change structure of sentence and remove redundant text -> This fulfills the requirements of the number type in HCL, even though it was not the most ideal solution.

File: report/4.Solution/GeneratedCppCode.tex [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+C++ does have an implementation of a string type in its standard library.
+However, the Arduino does not include this implementation.
+The Arduino uses its own string implementation that varies from the standard implementation.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Remove "That varies from the standard implementation" and add a "." after "implementation".
It is quite obvious that the custom implementation would differ from the standard implementation.

File: report/4.Solution/GeneratedCppCode.tex [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+Along with this, the string type shares a lot of functionality with that of lists.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Replace sentence with "The string type shares a lot of functionality with lists"

File: report/4.Solution/GeneratedCppCode.tex [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+Therefore the representation of strings in the generated C++ code, is based on the ConstList that is explained it the list section below.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Replace " that is explained it the list section below" -> ", which is explained below"

File: report/4.Solution/GeneratedCppCode.tex [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+\textbf{Lists}
+
+As for lists, C++ includes a list type in its standard library call vector. 
+However, this vector type is yet again not available in the default libraries provided by the Arduino.
+Furthermore the vector implementation is mutable.
+Since the lists in HCL is immutable, a simple implementation of a constant list was implemented.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: change "HCL is" -> "HCL are"

File: report/4.Solution/GeneratedCppCode.tex [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+One thing to be aware of, when using lists, is that the size cannot always be determined at compile time.
+This means that lists will usually do heap allocation. 
+With heap allocation comes the need to do proper memory handling, eg.
+clean up allocated memory in order to avoid memory leaks.
+
+This is done by utilizing the C++ concept of RAII\footnote{Resource acquisition is initialization} along with C++ smart pointers.
+All this really means, is that a pointer is implemented that keeps a count of the references to the memory pointed to.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Replace "All this really means, is that a" -> "Essentially"

File: report/4.Solution/GeneratedCppCode.tex [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+If the counter goes to zero, this means that there are no more references to memory, and the memory may be deallocated safely.
+
+It should be noted, that the current HCL compiler uses the std::shared\_ptr, which is not available in the Arduino standard library.
+The implementation should however be easily ported to the Arduino when needed.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: (Maybe) add a little bit of text explaining how the implementation can be ported.

File: report/4.Solution/GeneratedCppCode.tex [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+\textbf{Tuples}
+
+Tuples are represented as type defined structures in C++.
+However, HCL has a theoretical infinite amount of tuples, so obviously not all tuples can be generated.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "amount of tuples" -> "amount of nested tuples"

File: report/4.Solution/GeneratedCppCode.tex [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+This is the problem known as dangling references.
+The problem occurs when a closure is returned, and executed after a referenced variable has gone out of scope.
+The reference now points to some unallocated garbage memory, which leaves undefined behavior in the program.
+
+Although this feature would be nice to have in HCL, it also has to be deemed as undefined behavior for HCL when the target language is C++ and the closures are implemented using the method above.
+
+\textbf{Generics}
+
+As for generic types both used in functions and lists, C++ has a language concept known as templates.
+Templates are C++'s version of generics, and basically works by figuring out all the template variants used and then creating functions based this.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "based this" -> "based on these"

File: report/4.Solution/GeneratedCppCode.tex [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+Templates are not as powerful as modern generics known from for instance C\#, but thew will be enough for HCL, as HCL's generics are also not that powerful.
+One lacking feature of both, is the ability to create an instance of type T.
+
+\textbf{Naming of outputted types and variables}
+
+Aside from storing the types appropriately, a couple of other issues arose when doing the code generation.
+An interesting thing to mention is the naming changes from HCL to C++.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "naming changes" -> "identifier name changes"

File: report/4.Solution/GeneratedCppCode.tex [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+In HCL a lot of special chars are allowed for variable naming.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "variable" -> "identifier"

File: report/4.Solution/GeneratedCppCode.tex [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+\textbf{Naming of outputted types and variables}
+
+Aside from storing the types appropriately, a couple of other issues arose when doing the code generation.
+An interesting thing to mention is the naming changes from HCL to C++.
+In HCL a lot of special chars are allowed for variable naming. 
+This includes for instance "+" and "-" that is not allowed to use in C++, and even have a completely different meaning to the language.
+This was solved by hashing the name and prefixing it with for instance "IDT" for identifiers or "TPL" for tuples.
+
+Of course this solution leaves us vulnerable to hash code collisions.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: change "us" -> "the HCL compiler"

File: report/4.Solution/GeneratedCppCode.tex [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+We use a 32bit hash code, a long with the prefixes of types like "IDT" and "TPL", so it was accepted as a valid solution for now.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "We use a 32bit hash code, a long" -> "32bit hash codes are used, along"

File: report/4.Solution/GeneratedCppCode.tex (hunk 1 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 2 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 6 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 7 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 8 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 9 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed ish

File: report/4.Solution/GeneratedCppCode.tex (hunk 12 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex (hunk 16 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+Since C++ is the output language of the HCL compiler, all the language concepts of HCL have to be represented in valid C++ code.
+
+\textbf{Primitive types}
+The primitive types of HCL includes numbers, text and booleans. 
+Booleans are straightforward, as HCL's boolean behaves exactly like the C++ boolean type.
+This means that the regular C++ boolean type can be used as the output type of HCL booleans.
+
+As HCL aims for simplicity, only one number type is introduced. 
+This type needs to be able to store both integers and non-integers. 
+The simple solution is to turn all HCL numbers into C++'s double type.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "to turn all" -> "to format all"

File: report/4.Solution/GeneratedCppCode.tex [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+The solution with doubles is not ideal, but it will fulfill the requirements of the number type in HCL.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Cahnge structure of sentence and remove redundant text -> "This fulfills the requirements of the number type in HCL, even though it was not the most ideal solution"

File: report/4.Solution/GeneratedCppCode.tex [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+C++ does have an implementation of a string type in its standard library.
+However, the Arduino does not include this implementation.
+The Arduino uses its own string implementation, which varies from the standard implementation.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Remove "That varies from the standard implementation" and add a "." after "implementation".
It is quite obvious that the custom implementation would differ from the standard implementation.

File: report/4.Solution/GeneratedCppCode.tex [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+This means that lists will usually do heap allocation. 
+With heap allocation comes the need to do proper memory handling, eg.
+clean up allocated memory in order to avoid memory leaks.
+
+Memory handling is done by utilizing the C++ concept of RAII\footnote{Resource acquisition is initialization} along with C++ smart pointers.
+This means that a pointer is implemented, which keeps a count of the references to the memory pointed to.
+If the counter goes to zero, this means that there are no more references to memory, and the memory may be deallocated safely.
+
+It should be noted, that the current HCL compiler uses the std::shared\_ptr, which is not available in the Arduino standard library.
+The implementation should however be easily ported to the Arduino, however it is not complete as of the time of writing.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: (Maybe) add a little bit of text explaining how the implementation can be ported.

File: report/4.Solution/GeneratedCppCode.tex (hunk 11 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Fixed

File: report/4.Solution/GeneratedCppCode.tex [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+As for generic types, which is used in functions and lists, C++ has a language concept known as templates.
+Templates are C++'s version of generics, and works by determining all the template variants used and then creating functions according to these variants.
+Templates are not as powerful as modern generics, for example C\#'s generics, but they are enough for HCL, as HCL's generics are also not that powerful.
+One lacking feature of both, is the ability to create an instance of type T.
+
+\textbf{Naming of outputted types and variables}
+
+Aside from storing the types appropriately, there are a few other issues when generating code for C++
+An interesting thing to mention is the naming changes from HCL to C++.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "naming changes" -> "identifier name changes"

File: report/4.Solution/GeneratedCppCode.tex [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+In HCL a lot of special characters are allowed for variable naming.
```
Comment (jones1994) [inline report/4.Solution/GeneratedCppCode.tex]: Change "variable" -> "identifier"

File: report/4.Solution/GeneratedCppCode.tex (hunk 18 above)
Comment (nillas12) [reply report/4.Solution/GeneratedCppCode.tex]: "...It is not the ideal solution"

File: report/4.Solution/GeneratedCppCode.tex (hunk 19 above)
Comment (Lambcreek) [reply report/4.Solution/GeneratedCppCode.tex]: Hvad med at flette linje 18 og 19 sammen til "However, the Arduino does not include this implementation, but rather uses its own."

File: report/4.Solution/GeneratedCppCode.tex [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+However, the Arduino does not include this implementation.
+The Arduino uses its own string implementation, which varies from the standard implementation.
+
+As the string type shares a lot of functionality with that of lists, the representation of strings in the generated C++ code, is based on the ConstList.
+
+\textbf{Lists}
+
+For lists, C++ includes a list type in its standard library called vector. 
+This vector type is, like the string type, not available in the libraries provided by the Arduino.
+Furthermore the vector implementation is mutable and the lists in HCL are immutable, a simple implementation of a constant list is implemented instead.
```
Comment (Lambcreek) [inline report/4.Solution/GeneratedCppCode.tex]: Start sætningen med "Furthermore, since"

File: report/4.Solution/GeneratedCppCode.tex [hunk 24]
```diff
@@ -0,0 +1,86 @@
+% !TeX root = ../main.tex
+
+\subsection{Generated C++ code}
```
Comment (Lambcreek) [inline report/4.Solution/GeneratedCppCode.tex]: Skift alle de steder, hvor der står "C++" ud med "\texttt{C++}"
Det ser pænere ud, og det er det, som vi har gjort alle andre steder

File: report/4.Solution/GeneratedCppCode.tex (hunk 19 above)
Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: That works even better :)

File: report/4.Solution/GeneratedCppCode.tex (hunk 17 above)
Comment (technight) [reply report/4.Solution/GeneratedCppCode.tex]: Maybe turn is not the correct word, but i don't think format is correct either (the formating of the number remains the same, however the type is "turned" or changed to the double type). Any alternative suggestions - maybe "change all HCL number types into C++ doubles"? Althoug I still believe that turn can work here.

Comment (nillas12) [reply report/4.Solution/GeneratedCppCode.tex]: Convert, translate. Tho i think turn is fine

File: report/4.Solution/GeneratedCppCode.tex (hunk 18 above)
Comment (technight) [reply report/4.Solution/GeneratedCppCode.tex]: Not to be defensive, but I don't really understand the requested change.
First of all IMO both are perfectly valid English, with the exact same meaning.
Second of all the original sentence is shorter than the proposed sentence, so the part about removing redundant stuff seems strange.
Again, not being to defensive here, just genuinely puzzled why the proposed sentence is better? :smile:

File: report/4.Solution/GeneratedCppCode.tex (hunk 17 above)
Comment (technight) [reply report/4.Solution/GeneratedCppCode.tex]: Convert is perfect, tywm

File: report/4.Solution/GeneratedCppCode.tex (hunk 18 above)
Comment (technight) [reply report/4.Solution/GeneratedCppCode.tex]: Just fyi this one is the only open comment as far as I can see.

Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: Yea, that's a good point haha

Comment (jones1994) [reply report/4.Solution/GeneratedCppCode.tex]: that is a good point haha

**Pull request merged**